File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_generics.rs
//! Roundtrip coverage for `baml_sdk::generics` — generic classes (over `<int>`).
//!
//! The generic *instance method* path (`WrapperMethods::get_value` /
//! `get_value_or_marker`) is covered separately in
//! `customizable/test_generic.rs`; here we cover the
//! concretely-instantiated generic class round trips.

// Generic classes emit as `BamlValue`-bounded structs with pub fields,
// instantiated with struct-path turbofish (`Wrapper::<i64> { .. }`), and
// recursive fields box (`next: Option<std::boxed::Box<GenericLinkedList<T>>>`).
// The generated `Box` (the fixture's `Box<T>` class) shadows the prelude's
// `std::boxed::Box`, so the recursion boxes below are spelled in full.
use baml_sdk::generics::{
    Box, DifferingInstantiation, GenericBinaryTree, GenericLinkedList, NestedGenerics, Wrapper,
    round_trip_box_int, round_trip_differing_instantiation, round_trip_generic_binary_tree_int,
    round_trip_generic_linked_list_int, round_trip_nested_generics, round_trip_wrapper_int,
};

#[test]
fn test_generics_round_trip_wrapper_int() {
    let w = Wrapper::<i64> { value: 5 };
    assert_eq!(round_trip_wrapper_int(w.clone()).unwrap(), w);
}

#[test]
fn test_generics_round_trip_generic_linked_list_int() {
    let ll = GenericLinkedList::<i64> {
        value: 1,
        next: Some(std::boxed::Box::new(GenericLinkedList::<i64> {
            value: 2,
            next: None,
        })),
    };
    assert_eq!(round_trip_generic_linked_list_int(ll.clone()).unwrap(), ll);
}

#[test]
fn test_generics_round_trip_generic_binary_tree_int() {
    let t = GenericBinaryTree::<i64> {
        value: 1,
        left: None,
        right: None,
    };
    assert_eq!(round_trip_generic_binary_tree_int(t.clone()).unwrap(), t);
}

#[test]
fn test_generics_round_trip_box_int() {
    let b = Box::<i64> {
        value: 3,
        wrapped: Wrapper::<i64> { value: 4 },
    };
    assert_eq!(round_trip_box_int(b.clone()).unwrap(), b);
}

#[test]
fn test_generics_round_trip_nested_generics() {
    let n = NestedGenerics {
        ww: Wrapper::<Wrapper<i64>> {
            value: Wrapper::<i64> { value: 1 },
        },
        // python spells this `Wrapper[list]`; the field type is
        // `Wrapper<int[]>`, i.e. `Wrapper<Vec<i64>>`.
        wl: Wrapper::<Vec<i64>> { value: vec![1, 2] },
        wr: Wrapper::<GenericLinkedList<i64>> {
            value: GenericLinkedList::<i64> {
                value: 9,
                next: None,
            },
        },
    };
    assert_eq!(round_trip_nested_generics(n.clone()).unwrap(), n);
}

#[test]
fn test_generics_round_trip_differing_instantiation() {
    let d = DifferingInstantiation {
        list: GenericLinkedList::<Wrapper<i64>> {
            value: Wrapper::<i64> { value: 1 },
            next: None,
        },
    };
    assert_eq!(round_trip_differing_instantiation(d.clone()).unwrap(), d);
}

File: baml_language/sdk_tests/crates/java/type_shapes/customizable/roundtrip_tests/TestGenerics.java
// Roundtrip coverage for `baml_sdk.generics` — generic classes (over
// `<int>`).
//
// The generic *instance method* path (`WrapperMethods.get_value` /
// `get_value_or_marker`) is covered separately in top-level
// `TestGeneric.java`, mirroring `customizable/test_generic.py`; here we
// cover the concretely-instantiated generic class round trips.
//
// Port of python_pydantic2/type_shapes/customizable/roundtrip_tests/
// test_generics.py — same test names, cases, inputs, assertions.
//
// java-port note: the conventions doc pins `new Wrapper<Long>(5L)` /
// `Wrapper<Long> w = ...` as the generic-class construction shape but
// leaves "explicit type binding" as a trailing-type-args-parameter TBD (see
// the "Generic function/method (explicit)" row in
// ref-java-state-of-completeness.md: "Java has no `_types=` kwarg or
// subscript; shape TBD"). This port uses the plain generic constructor form
// exactly as given in the task's own example, with no trailing type-args
// parameter. How the hidden `BamlType[]` type-args side-channel mentioned
// for "Generic explicitly/implicitly reified" values gets populated from an
// ordinary `new Wrapper<Long>(...)` call (Java generics are erased at
// runtime) is unresolved and needs a real codegen decision.
package roundtrip_tests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import baml_sdk.generics.Box;
import baml_sdk.generics.DifferingInstantiation;
import baml_sdk.generics.Fns;
import baml_sdk.generics.GenericBinaryTree;
import baml_sdk.generics.GenericLinkedList;
import baml_sdk.generics.NestedGenerics;
import baml_sdk.generics.Wrapper;
import java.util.List;
import org.junit.jupiter.api.Test;

class TestGenerics {

    @Test
    void test_generics_round_trip_wrapper_int() {
        Wrapper<Long> w = new Wrapper<>(5L);
        assertEquals(w, Fns.round_trip_wrapper_int(w));
    }

    @Test
    void test_generics_round_trip_generic_linked_list_int() {
        GenericLinkedList<Long> ll =
                new GenericLinkedList<>(1L, new GenericLinkedList<>(2L, null));
        assertEquals(ll, Fns.round_trip_generic_linked_list_int(ll));
    }

    @Test
    void test_generics_round_trip_generic_binary_tree_int() {
        GenericBinaryTree<Long> t = new GenericBinaryTree<>(1L, null, null);
        assertEquals(t, Fns.round_trip_generic_binary_tree_int(t));
    }

    @Test
    void test_generics_round_trip_box_int() {
        Box<Long> b = new Box<>(3L, new Wrapper<>(4L));
        assertEquals(b, Fns.round_trip_box_int(b));
    }

    @Test
    void test_generics_round_trip_nested_generics() {
        NestedGenerics n =
                new NestedGenerics(
                        new Wrapper<Wrapper<Long>>(new Wrapper<>(1L)),
                        new Wrapper<List<Long>>(List.of(1L, 2L)),
                        new Wrapper<GenericLinkedList<Long>>(new GenericLinkedList<>(9L, null)));
        assertEquals(n, Fns.round_trip_nested_generics(n));
    }

    @Test
    void test_generics_round_trip_differing_instantiation() {
        DifferingInstantiation d =
                new DifferingInstantiation(
                        new GenericLinkedList<Wrapper<Long>>(new Wrapper<>(1L), null));
        assertEquals(d, Fns.round_trip_differing_instantiation(d));
    }
}
